File: app/services/workbench/candidate_rank.py
"""Stable ranking helpers for workbench semantic groups."""

from __future__ import annotations

from app.domain.lexicon.ranking import search_result_sort_key


def relation_index(rows: list[dict]) -> dict[str, tuple[int, str | None]]:
    return {
        str(row.get("char")): (index, row.get("source"))
        for index, row in enumerate(rows)
        if row.get("char")
    }


def candidate_sort_key(candidate) -> tuple:
    return (
        candidate.source_rank,
        candidate.literal,
        candidate.jyutping,
    )


def sound_candidate_sort_key(candidate) -> tuple:
    """sound_only uses canonical search ranking with stable source tie-breaks."""
    return (
        *search_result_sort_key({"char": candidate.literal, "jyutping": candidate.jyutping}),
        candidate.source_rank,
        candidate.code,
    )


__all__ = ["candidate_sort_key", "relation_index", "sound_candidate_sort_key"]

File: client/src/workbench/candidate-rank.ts
import { compareSearchResults } from '../db/ranking.ts';
import type { WorkbenchCandidate } from './contracts.ts';

export interface RelationRank {
  rank: number;
  source?: string;
}

export function relationIndex(rows: Array<{ char: string; source?: string }>): Map<string, RelationRank> {
  return new Map(rows.filter((row) => row.char).map((row, rank) => [row.char, { rank, source: row.source }]));
}

/** sound_only follows the canonical frequency ranking, not the source page offset. */
export function compareSoundOnlyCandidates(a: WorkbenchCandidate, b: WorkbenchCandidate): number {
  const ranked = compareSearchResults(
    { char: a.literal, jyutping: a.jyutping },
    { char: b.literal, jyutping: b.jyutping },
  );
  return ranked
    || a.sourceRank - b.sourceRank
    || a.code.localeCompare(b.code);
}
